tests/websocket.c

```diff
@@ -8,319 +8,537 @@
+
+    if(event != EV_READ) {
+        return;
+    }
+
+    /* read message */
+    ret = evbuffer_read(wt->rbuffer, fd, 65536);
+    wt->debug("evbuffer_read() returned %d; wt->state=%d. wt->rbuffer:\n", ret, wt->state);
+    evbuffer_debug_dump(wt, wt->rbuffer);
+    if (ret == 0) {
+        wt->debug("We didn't read anything from the socket...\n");
+        return;
+    }
+
+    while(1) {
+        switch (wt->state) {
+        case WS_SENT_HANDSHAKE: { /* waiting for handshake response */
+            size_t avail_sz = evbuffer_get_length(wt->rbuffer);
+            char *tmp = calloc(avail_sz, 1);
+            wt->debug("avail_sz from rbuffer = %lu\n", avail_sz);
+            evbuffer_remove(wt->rbuffer, tmp, avail_sz); /* copy into `tmp` */
+            wt->debug("Giving %lu bytes to http-parser\n", avail_sz);
+            int nparsed = http_parser_execute(&wt->parser, &wt->settings, tmp, avail_sz);
+            wt->debug("http-parser returned %d\n", nparsed);
+            if (nparsed != (int)avail_sz) { // put back what we didn't read
+                wt->debug("re-attach (prepend) %lu byte%c\n", avail_sz - nparsed,
+                          avail_sz - nparsed > 1 ? 's' : ' ');
+                evbuffer_prepend(wt->rbuffer, tmp + nparsed, avail_sz - nparsed);
+            }
+            free(tmp);
+            if (wt->state == WS_SENT_HANDSHAKE &&  /* haven't encountered end of response yet */
+                wt->parser.upgrade && nparsed != (int)avail_sz) {
+                wt->debug("UPGRADE *and* we have some data left\n");
+                continue;
+            } else if (wt->state == WS_RECEIVED_HANDSHAKE) { /* we have the full response */
+                evbuffer_drain(wt->rbuffer, evbuffer_get_length(wt->rbuffer));
+            }
+            return;
+        }
+
+        case WS_SENT_FRAME: { /* waiting for frame response */
+            wt->debug("We're in WS_SENT_FRAME, just read a frame response. wt->rbuffer:\n");
+            evbuffer_debug_dump(wt, wt->rbuffer);
+            uint8_t flag_opcodes, payload_len;
+            if (evbuffer_get_length(wt->rbuffer) < 2) { /* not enough data */
+                wait_for_possible_read(wt);
+                return;
+            }
+            evbuffer_remove(wt->rbuffer, &flag_opcodes, 1); /* remove flags & opcode */
+            evbuffer_remove(wt->rbuffer, &payload_len, 1);  /* remove length */
+            evbuffer_drain(wt->rbuffer, (size_t)payload_len); /* remove payload itself */
+            process_message(wt, payload_len);
+
+            if (evbuffer_get_length(wt->rbuffer) == 0) { /* consumed everything */
+                if (wt->msg_received < wt->msg_target) { /* let's write again */
+                    wt->debug("our turn to write again\n");
+                    wt->state = WS_RECEIVED_HANDSHAKE;
+                    ws_enqueue_frame(wt);
+                } /* otherwise, we're done */
+                return;
+            } else {
```

Comment:
when would there still be data after reading the full payload with `evbuffer_drain`?

Reply:
It won't happen with the current messages we send, but if we start sending more than one at a time then we'll have the code for it. If we remove this it will make this part of the code dependent on the Redis commands, in a non-obvious way.